edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/DeviceScopeIdentitiesCache.cs

```diff
@@ -57,22 +64,38 @@ public sealed class DeviceScopeIdentitiesCache : IDeviceScopeIdentitiesCache
             IServiceIdentityHierarchy serviceIdentityHierarchy,
             IServiceProxy serviceProxy,
             IKeyValueStore<string, string> encryptedStorage,
-            TimeSpan refreshRate)
+            TimeSpan refreshRate,
+            TimeSpan refreshDelay)
         {
             Preconditions.CheckNotNull(serviceProxy, nameof(serviceProxy));
             Preconditions.CheckNotNull(encryptedStorage, nameof(encryptedStorage));
             Preconditions.CheckNotNull(serviceIdentityHierarchy, nameof(serviceIdentityHierarchy));
             IDictionary<string, StoredServiceIdentity> cache = await ReadCacheFromStore(encryptedStorage);
-            var deviceScopeIdentitiesCache = new DeviceScopeIdentitiesCache(serviceIdentityHierarchy, serviceProxy, encryptedStorage, cache, refreshRate);
+            var deviceScopeIdentitiesCache = new DeviceScopeIdentitiesCache(serviceIdentityHierarchy, serviceProxy, encryptedStorage, cache, refreshRate, refreshDelay);
             Events.Created();
             return deviceScopeIdentitiesCache;
         }
 
         public void InitiateCacheRefresh()
         {
             Events.ReceivedRequestToRefreshCache();
-            this.refreshCacheCompleteSignal.Reset();
-            this.refreshCacheSignal.Set();
+
+            TimeSpan timeSinceLastRefresh = DateTime.UtcNow - this.cacheLastRefreshTime;
+
+            // Only refresh the cache if we haven't done so recently
+            if (timeSinceLastRefresh > this.refreshDelay)
+            {
+                this.refreshCacheCompleteSignal.Reset();
+                this.refreshCacheSignal.Set();
+
+                // Update the cache refresh timestamp
+                this.cacheLastRefreshTime = DateTime.UtcNow;
+            }
+            else
+            {
+                Events.SkipRefreshCache(this.cacheLastRefreshTime, this.refreshDelay);
+                this.refreshCacheCompleteSignal.Set();
```

Comment:
Current we return refreshCacheCompleteSignal.Set() immediately, this may cause problems.  For example, there are 2 requests to call to initiateCacheRefresh.  The 1st request successfully goes into line 88 and set refresthCacheSignal, then RefreshCache method will call serviceProxy to get identities.  Then 2nd request comes in and since it is still within refresh delay (2 mins), it will call refreshCacheCompleteSignal.Set() in line97 which will release both 1st and 2nd request WaitForCacheRefresh calls.  I think we want both 1st and 2nd requests should wait for 1st request identities result comes back, right?

should we just remove line 97 since this.refreshCacheCompleteSignal.Set() should be called once 1st request is finished?

Reply:
Do you have test to verify this case?